Sources/NIOEmbedded/AsyncTestingChannel.swift

```diff
@@ -0,0 +1,508 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+import NIOConcurrencyHelpers
+import NIOCore
+
+/// `NIOAsyncTestingChannel` is a `Channel` implementation that does no
+/// actual IO but that does have proper eventing mechanism, albeit one that users can
+/// control. The prime use-case for `NIOAsyncTestingChannel` is in unit tests when you
+/// want to feed the inbound events and check the outbound events manually.
+///
+/// Please remember to call `finish()` when you are no longer using this
+/// `NIOAsyncTestingChannel`.
+///
+/// To feed events through an `NIOAsyncTestingChannel`'s `ChannelPipeline` use
+/// `NIOAsyncTestingChannel.writeInbound` which accepts data of any type. It will then
+/// forward that data through the `ChannelPipeline` and the subsequent
+/// `ChannelInboundHandler` will receive it through the usual `channelRead`
+/// event. The user is responsible for making sure the first
+/// `ChannelInboundHandler` expects data of that type.
+///
+/// `NIOAsyncTestingChannel` automatically collects arriving outbound data and makes it
+/// available one-by-one through `readOutbound`.
+///
+/// - note: `NIOAsyncTestingChannel` is currently only compatible with
+///   `NIOAsyncTestingEventLoop`s and cannot be used with `SelectableEventLoop`s from
+///   for example `MultiThreadedEventLoopGroup`.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+public final class NIOAsyncTestingChannel: Channel {
+    /// `LeftOverState` represents any left-over inbound, outbound, and pending outbound events that hit the
+    /// `NIOAsyncTestingChannel` and were not consumed when `finish` was called on the `NIOAsyncTestingChannel`.
+    ///
+    /// `NIOAsyncTestingChannel` is most useful in testing and usually in unit tests, you want to consume all inbound and
+    /// outbound data to verify they are what you expect. Therefore, when you `finish` an `NIOAsyncTestingChannel` it will
+    /// return if it's either `.clean` (no left overs) or that it has `.leftOvers`.
+    public enum LeftOverState {
+        /// The `NIOAsyncTestingChannel` is clean, ie. no inbound, outbound, or pending outbound data left on `finish`.
+        case clean
+
+        /// The `NIOAsyncTestingChannel` has inbound, outbound, or pending outbound data left on `finish`.
+        case leftOvers(inbound: CircularBuffer<NIOAny>, outbound: CircularBuffer<NIOAny>, pendingOutbound: [NIOAny])
```

Comment:
Do we want to continue using `CircularBuffer` in new public API or should we start switching to `Deque` where possible? 

(I know these come from the channel core and we probably don't want to incur the additional allocation cost from copying them over but I'm still interested in the answer to the general case)

Reply:
Yeah that's a good question. In this instance we should leave it due to the existing use in the embedded channel, but I think that we should be using `Deque` in new API if we don't have a good reason not to.